python/lib/communication/dmod/communication/client.py

```diff
@@ -58,31 +46,87 @@ def get_or_create_eventloop() -> AbstractEventLoop:
         raise
 
 
-class AbstractClient(ABC):
+class TransportLayerClient(ABC):
     """
-    Abstract client capable of securely communicating with a server at some endpoint.
+    Abstract client capable of communicating with a server at some endpoint.
+
+    Abstract client for interacting with a service at the OSI transport layer.  It provides an interface for sending
+    data to accept data and send this data to a server at some endpoint.  The interface function for this behavior
+    supports optionally waiting for and returning a raw data response.  Alternatively, the type provides a function for
+    receiving a response from the server independently.
 
-    Abstract client with an interface for securely sending data to a server at some endpoint.  The interface function
-    for this behavior supports optionally waiting for and returning a raw response.  Alternatively, the type provides an
-    interface for receiving a response from the server independently.
+    Instances are capable of securing communications using an ::class:`SSLContext`.  A customized context or default
+    context can be created, depending on the parameters passed during init.
     """
-    def __init__(self, endpoint_uri: str, *args, **kwargs):
+
+    @classmethod
+    @abstractmethod
+    def get_endpoint_protocol_str(cls, use_secure_connection: bool = True) -> str:
+        """
+        Get the protocol substring portion for valid connection URI strings for an instance of this class.
+
+        Parameters
+        ----------
+        use_secure_connection : bool
+            Whether to get the protocol substring applicable for secure connections (``True`` by default).
+
+        Returns
+        -------
+        str
+            The protocol substring portion for valid connection URI strings for an instance of this class.
+        """
+        pass
+
+    def __init__(self, endpoint_host: str, endpoint_port: Union[int, str], endpoint_path: Optional[str] = None,
```

Comment:
This feels like it is doing a lot, but I think it is alright. Thanks for reworking this @robertbartel! My only comment is, thoughts on making all arguments required keyword arguments? So the signature would change to:

```python
      def __init__(
        self,
        *,
        endpoint_host: str,
        endpoint_port: Union[int, str],
        endpoint_path: Optional[str] = None,
        cafile: Optional[Path] = None,
        capath: Optional[Path] = None,
        use_default_context: bool = False,
        **kwargs
    ):
```

Reply:
Sure, I've just made and pushed that change.